src/rpcrequest-payload.ts

```diff
@@ -116,11 +135,81 @@ class RpcRequestPayload {
       param.scale = type.resolveScale(parameter);
     }
 
+    if (parameter.collation) {
+      param.collation = parameter.collation;
+    }
+
     type.writeTypeInfo(buffer, param, this.options);
-    type.writeParameterData(buffer, param, this.options, () => {
+    if (writeValue) {
+      type.writeParameterData(buffer, param, this.options, () => {
+        cb();
+      });
+    } else {
       cb();
+    }
+  }
+
+  _writeEncryptedParameter(parameter: Parameter, buffer: WritableTrackingBuffer, cb: () => void) {
+    const encryptedParam = {
+      value: parameter.encryptedVal,
+      type: VarBinary,
+      output: parameter.output,
+      name: parameter.name,
+      forceEncrypt: parameter.forceEncrypt
+    };
+
+    this._writeParameterData(encryptedParam, buffer, true, () => {
+      this._writeParameterData(parameter, buffer, false, () => {
+        this._writeEncryptionMetadata(parameter.cryptoMetadata, buffer, () => {
+          cb();
+        });
+      });
     });
   }
+
+  _encryptParameters(parameters: Parameter[], callback: (parameters: Parameter[]) => void) {
+    if (this.options.serverSupportsColumnEncryption === true) {
+      const promises: Promise<void>[] = [];
+
+      for (let i = 0, len = parameters.length; i < len; i++) {
+        const type = parameters[i].type;
+        if (parameters[i].cryptoMetadata && parameters[i].value) {
+          if (!type.toBuffer) {
+            throw new Error(`Column encryption error. Cannot convert type ${type.name} to buffer.`);
+          }
+          const plainTextBuffer = type.toBuffer(parameters[i], this.options);
+          if (plainTextBuffer) {
+            promises.push(encryptWithKey(plainTextBuffer, parameters[i].cryptoMetadata as CryptoMetadata, this.options).then((encryptedValue: Buffer) => {
+              parameters[i].encryptedVal = encryptedValue;
+            }));
+          }
+        }
+      }
+
+      Promise.all(promises).then(() => callback(parameters));
```

Comment:
@fredwes Is there a reason why `encryptWithKey` is wrapped in a promise? If `serverSupportsColumnEncryption = true`, then before the call back is called (on line 189), it would have to wait for all the parameters to be encrypted, and there's nothing else to do in this function in the meantime as it's being encrypted. Wouldn't that just make this process synchronous then?

Reply:
@IanChokS It waits for all parameters to be encrypted before continuing. We saw a significant performance improvement this way. Especially when parameters use different master/column encryption keys (since fetching the keys can happen in parallel this way). It may be worth testing this again with the recent generator refactors though